dataset/src/main/scala/org/apache/spark/sql/FramelessInternals.scala

```diff
@@ -34,4 +34,6 @@ object FramelessInternals {
   def ofRows(sparkSession: SparkSession, logicalPlan: LogicalPlan): DataFrame = {
     Dataset.ofRows(sparkSession, logicalPlan)
   }
+
+  type PublicUserDefinedType[A >: Null] = UserDefinedType[A]
```

Comment:
What's the motivation to not inline this type all the way? (instead of introducing `frameless.udt.Udt` and `FramelessInternals.PublicUserDefinedType`)

Reply:
I defined org.apache.spark.sql.FramelessInternals.PublicUserDefinedType to have a public type for Spark's UserDefinedType (which is private[spark]) that I can reuse in other packages.

I defined frameless.udt.Udt to avoid the user wanting to define a new Udt to import org.apache.spark.sql.FramelessInternals in their code.

Reply:
I see. But with a type alias you are not actually enforcing anything, as in i think an `implicit val pudt: PublicUserDefinedType[Foo]` would be picked up by the compiler as a perfectly valid instance of `frameless.udt.Udt[Foo]`. If you want to prevent that I think you should use a proper new type instead (`case class Udt[A](...) extends AnyVal`).

Reply:
Yes indeed, my goal was more to avoid extending FramelessInternals.PublicUserDefinedType when defining a new Udt like in https://github.com/typelevel/frameless/pull/166/files#diff-29139c6b2def8a3ae909edff01b315c9R27, but you're right as Udt are private in Spark for now, we may should discourage user using it directly (apart from already defined Udt like in spark-ml). So in the end I could remove frameless.udt.Udt and only use FramelessInternals.PublicUserDefinedType in TypedEncoders etc.

Reply:
I don't have a strong opinion on this, I just wanted to point out that type alias enforce less restrictions than value classes :)

Reply:
@OlivierBlanvillain modification pushed :) I think this is clearer this way that UserDefinedType is a Spark structure (not Frameless).